axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorder/SaleOrderLineTreeServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+  }
+
+  @Override
+  @Transactional
+  public SaleOrderLineTree fillTheFields(SaleOrderLineTree saleOrderLineTree) {
+    Product product = saleOrderLineTree.getProduct();
+    saleOrderLineTree.setTitle(product.getName());
+    saleOrderLineTree.setUnit(product.getUnit());
+    saleOrderLineTree.setDescription(product.getDescription());
+    saleOrderLineTree.setUnitCost(product.getCostPrice());
+    saleOrderLineTree.setUnitPrice(product.getSalePrice());
+    saleOrderLineTree.setTotalCost(
+        product.getCostPrice().multiply(saleOrderLineTree.getQuantity()));
+    saleOrderLineTree.setTotalPrice(
+        product.getSalePrice().multiply(saleOrderLineTree.getQuantity()));
+
+    if (saleOrderLineTree.getUnitPrice().compareTo(BigDecimal.ZERO) != 0) {
+      saleOrderLineTree.setMarginRate(
+          new BigDecimal(1)
+              .min(
+                  saleOrderLineTree
+                      .getUnitCost()
+                      .divide(saleOrderLineTree.getUnitPrice(), 2, RoundingMode.HALF_UP)));
+    }
+    return saleOrderLineTree;
+  }
+
+  @Override
+  public SaleOrderLineTree updateTheFields(SaleOrderLineTree saleOrderLineTree) {
+
+    saleOrderLineTree.setUnit(saleOrderLineTree.getUnit());
+
+    saleOrderLineTree.setTotalCost(
+        saleOrderLineTree.getUnitCost().multiply(saleOrderLineTree.getQuantity()));
+    saleOrderLineTree.setTotalPrice(
+        saleOrderLineTree.getUnitPrice().multiply(saleOrderLineTree.getQuantity()));
+    if (saleOrderLineTree.getUnitPrice().compareTo(BigDecimal.ZERO) != 0) {
+      saleOrderLineTree.setMarginRate(
+          new BigDecimal(1)
+              .min(
+                  saleOrderLineTree
+                      .getUnitCost()
+                      .divide(saleOrderLineTree.getUnitPrice(), 2, RoundingMode.HALF_UP)));
+    }
+    return saleOrderLineTree;
+  }
+
+  @Override
+  public SaleOrderLineTree updateUnitPrice(SaleOrderLineTree saleOrderLineTree) {
+    if (new BigDecimal(1).min(saleOrderLineTree.getMarginRate()).compareTo(BigDecimal.ZERO) != 0) {
+      saleOrderLineTree.setUnitPrice(
+          saleOrderLineTree
+              .getUnitCost()
+              .divide(
+                  new BigDecimal(1).min(saleOrderLineTree.getMarginRate()),
+                  2,
+                  RoundingMode.HALF_UP));
+    }
+    saleOrderLineTree.setTotalPrice(
+        saleOrderLineTree.getUnitPrice().multiply(saleOrderLineTree.getQuantity()));
+    return saleOrderLineTree;
+  }
+
+  @Override
+  @Transactional
+  public void removeSubElement(SaleOrderLineTree saleOrderLineTree) throws AxelorAlertException {
+
+    removeElement(saleOrderLineTree);
+  }
+
+  public void removeElement(SaleOrderLineTree saleOrderLineTree) {
+
+    List<SaleOrderLineTree> saleOrderLineTrees =
+        saleOrderLineTreeRepository
+            .all()
+            .filter("self.parent=?1", saleOrderLineTree.getId())
+            .fetch();
+
+    if (CollectionUtils.isEmpty(saleOrderLineTrees)) {
```

Comment:
```suggestion
    if (ObjectUtils.isEmpty(saleOrderLineTrees)) {
```

Use axelor libs in favor of third-party libs. Please review the other changed files too.